File: sdk/java/src/main/java/com/agentquay/JsonSchemaGenerator.java
package com.agentquay;

import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.databind.BeanDescription;
import com.fasterxml.jackson.databind.JavaType;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.introspect.BeanPropertyDefinition;

import java.lang.reflect.Method;
import java.lang.reflect.Parameter;
import java.lang.reflect.ParameterizedType;
import java.lang.reflect.Type;
import java.util.ArrayList;
import java.util.HashSet;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.Set;

/**
 * 从方法签名与类型注解生成 JSON Schema（设计文档 §4.4 "Jackson 生成 JSON Schema"）。
 *
 * <p>基于 Jackson 的类型系统（JavaType + BeanDescription），支持：
 * 基本类型、String、enum、List/Set/数组、Map、Optional、嵌套 POJO，
 * 以及 Jackson 注解（@JsonProperty 重命名 / @JsonIgnore 忽略）。
 * 方法参数名依赖编译期 {@code -parameters} 保留。
 */
public final class JsonSchemaGenerator {

    private final ObjectMapper mapper;

    public JsonSchemaGenerator() {
        this.mapper = new ObjectMapper();
    }

    /**
     * 生成方法参数的 inputSchema：{@code {"type":"object","properties":{...},"required":[...]}}。
     */
    public Map<String, Object> paramSchema(Method method) {
        Map<String, Object> properties = new LinkedHashMap<>();
        List<String> required = new ArrayList<>();

        int index = 0;
        for (Parameter param : method.getParameters()) {
            AgentParam ap = param.getAnnotation(AgentParam.class);
            boolean markedRequired = ap == null || ap.required();
            boolean optionalType = isOptional(param.getParameterizedType());

            String propName = param.getName();
            // 编译器未保留参数名时回退 "argN" 并警告（pom 已开启 -parameters）
            if (propName == null || propName.startsWith("arg")) {
                propName = "arg" + index;
            }
            index++;

            Map<String, Object> schema = typeSchema(param.getParameterizedType());
            if (ap != null && !ap.description().isEmpty()) {
                schema.put("description", ap.description());
            }
            properties.put(propName, schema);

            if (markedRequired && !optionalType) {
                required.add(propName);
            }
        }

        Map<String, Object> root = new LinkedHashMap<>();
        root.put("type", "object");
        root.put("properties", properties);
        if (!required.isEmpty()) {
            root.put("required", required);
        }
        return root;
    }

    /** 类型 → JSON Schema（Optional 类型剥壳并标记可空）。 */
    public Map<String, Object> typeSchema(Type type) {
        return typeSchema(type, new HashSet<>());
    }

    private Map<String, Object> typeSchema(Type type, Set<Class<?>> seen) {
        if (type instanceof ParameterizedType) {
            ParameterizedType pt = (ParameterizedType) type;
            Class<?> raw = (Class<?>) pt.getRawType();
            Type[] args = pt.getActualTypeArguments();

            if (raw == Optional.class) {
                Map<String, Object> inner = typeSchema(args[0], seen);
                // Optional → 可空
                if (inner.containsKey("type")) {
                    Object t = inner.get("type");
                    if (t instanceof String) {
                        inner.put("type", java.util.Arrays.asList(t, "null"));
                    }
                } else {
                    inner.put("nullable", true);
                }
                return inner;
            }
            if (List.class.isAssignableFrom(raw) || Set.class.isAssignableFrom(raw)
                    || Iterable.class.isAssignableFrom(raw)) {
                Map<String, Object> schema = new LinkedHashMap<>();
                schema.put("type", "array");
                if (Set.class.isAssignableFrom(raw)) {
                    schema.put("uniqueItems", true);
                }
                schema.put("items", args.length > 0 ? typeSchema(args[0], seen) : emptySchema());
                return schema;
            }
            if (Map.class.isAssignableFrom(raw)) {
                Map<String, Object> schema = new LinkedHashMap<>();
                schema.put("type", "object");
                schema.put("additionalProperties",
                        args.length > 1 ? typeSchema(args[1], seen) : emptySchema());
                return schema;
            }
        }

        Class<?> raw = rawClass(type);
        if (raw == null) {
            return emptySchema();
        }
        if (raw == String.class || raw == Character.class || raw == char.class) {
            return mapOf("type", "string");
        }
        if (raw == Integer.class || raw == int.class || raw == Long.class || raw == long.class
                || raw == Short.class || raw == short.class || raw == Byte.class || raw == byte.class) {
            return mapOf("type", "integer");
        }
        if (raw == Double.class || raw == double.class || raw == Float.class || raw == float.class) {
            return mapOf("type", "number");
        }
        if (raw == Boolean.class || raw == boolean.class) {
            return mapOf("type", "boolean");
        }
        if (raw.isEnum()) {
            List<Object> values = new ArrayList<>();
            for (Object c : raw.getEnumConstants()) {
                values.add(c instanceof Enum ? ((Enum<?>) c).name() : String.valueOf(c));
            }
            Map<String, Object> schema = new LinkedHashMap<>();
            schema.put("enum", values);
            return schema;
        }
        if (raw.isArray()) {
            Map<String, Object> schema = new LinkedHashMap<>();
            schema.put("type", "array");
            schema.put("items", typeSchema(raw.getComponentType(), seen));
            return schema;
        }
        // 嵌套 POJO
        return beanSchema(raw, seen);
    }

    /** 用 Jackson BeanDescription 内省 POJO 属性。 */
    private Map<String, Object> beanSchema(Class<?> beanClass, Set<Class<?>> seen) {
        Map<String, Object> schema = new LinkedHashMap<>();
        schema.put("type", "object");
        Map<String, Object> properties = new LinkedHashMap<>();
        List<String> required = new ArrayList<>();

        if (seen.contains(beanClass)) {
            schema.put("properties", properties);
            return schema; // 循环引用：截断
        }
        seen.add(beanClass);
        try {
            JavaType javaType = mapper.getTypeFactory().constructType(beanClass);
            BeanDescription desc = mapper.getSerializationConfig().introspect(javaType);
            for (BeanPropertyDefinition prop : desc.findProperties()) {
                if (prop.getPrimaryMember() != null
                        && prop.getPrimaryMember().getAnnotation(JsonIgnore.class) != null) {
                    continue;
                }
                String name = prop.getName();
                if (prop.getPrimaryMember() != null) {
                    JsonProperty jp = prop.getPrimaryMember().getAnnotation(JsonProperty.class);
                    if (jp != null && !jp.value().isEmpty()) {
                        name = jp.value();
                    }
                }
                Map<String, Object> propSchema = typeSchema(reflectType(prop.getPrimaryType()), seen);
                if (prop.isRequired()) {
                    required.add(name);
                }
                properties.put(name, propSchema);
            }
        } finally {
            seen.remove(beanClass);
        }
        schema.put("properties", properties);
        if (!required.isEmpty()) {
            schema.put("required", required);
        }
        return schema;
    }

    private static boolean isOptional(Type type) {
        return type instanceof ParameterizedType
                && ((ParameterizedType) type).getRawType() == Optional.class;
    }

    /**
     * Jackson JavaType → java.lang.reflect.Type（尽力保留 List/Map/Optional 泛型）。
     * BeanDescription 的属性类型是 JavaType，而 typeSchema 基于 reflect 类型系统。
     */
    private static Type reflectType(JavaType jt) {
        if (jt == null) {
            return null;
        }
        Class<?> raw = jt.getRawClass();
        if (List.class.isAssignableFrom(raw) && jt.getContentType() != null) {
            return new ParameterizedTypeImpl(List.class,
                    new Type[]{reflectType(jt.getContentType())});
        }
        if (Map.class.isAssignableFrom(raw) && jt.getKeyType() != null
                && jt.getContentType() != null) {
            return new ParameterizedTypeImpl(Map.class,
                    new Type[]{reflectType(jt.getKeyType()), reflectType(jt.getContentType())});
        }
        if (Optional.class.isAssignableFrom(raw) && jt.getContentType() != null) {
            return new ParameterizedTypeImpl(Optional.class,
                    new Type[]{reflectType(jt.getContentType())});
        }
        return raw;
    }

    /** 最小 ParameterizedType 实现（JDK 无公开构造）。 */
    private static final class ParameterizedTypeImpl implements ParameterizedType {
        private final Class<?> raw;
        private final Type[] args;

        ParameterizedTypeImpl(Class<?> raw, Type[] args) {
            this.raw = raw;
            this.args = args;
        }

        @Override
        public Type[] getActualTypeArguments() {
            return args;
        }

        @Override
        public Type getRawType() {
            return raw;
        }

        @Override
        public Type getOwnerType() {
            return null;
        }
    }

    private static Class<?> rawClass(Type type) {
        if (type instanceof Class) {
            return (Class<?>) type;
        }
        if (type instanceof ParameterizedType) {
            Type raw = ((ParameterizedType) type).getRawType();
            return raw instanceof Class ? (Class<?>) raw : null;
        }
        return null;
    }

    private static Map<String, Object> mapOf(String k, Object v) {
        Map<String, Object> m = new LinkedHashMap<>();
        m.put(k, v);
        return m;
    }

    private static Map<String, Object> emptySchema() {
        return new LinkedHashMap<>();
    }
}

File: sdk/dotnet/src/AgentQuay.Sdk/JsonSchemaGenerator.cs
using System.Collections;
using System.Reflection;
using System.Text.Json.Nodes;
using System.Text.Json.Serialization;

namespace AgentQuay;

/// <summary>
/// 从方法签名与类型注解生成 JSON Schema（设计文档 §4.2 "System.Text.Json 生成 JSON Schema"）。
///
/// <para>支持：基本类型、string、char、Guid、DateTime、enum、数组、List/Set/集合、
/// IDictionary、可空类型（T? / Nullable&lt;T&gt;）、嵌套 POJO，以及
/// System.Text.Json 注解（[JsonPropertyName] 重命名 / [JsonIgnore] 忽略）。</para>
/// </summary>
public sealed class JsonSchemaGenerator
{
    /// <summary>
    /// 生成方法参数的 inputSchema：{"type":"object","properties":{...},"required":[...]}。
    /// </summary>
    /// <remarks>CancellationToken 参数不进入 schema（SDK 调用时自动注入 None）。</remarks>
    public JsonObject ParamSchema(MethodInfo method)
    {
        var properties = new JsonObject();
        var required = new List<string>();

        int index = 0;
        foreach (ParameterInfo param in method.GetParameters())
        {
            if (param.ParameterType == typeof(CancellationToken) ||
                param.ParameterType == typeof(System.Threading.CancellationToken?))
            {
                continue; // 基础设施参数：不入 schema
            }
            var ap = param.GetCustomAttribute<AgentParamAttribute>();
            bool markedRequired = ap?.Required ?? true;
            bool explicitOptional = IsNullableLike(param);

            string propName = param.Name ?? "arg" + index;
            // 编译器生成参数名 / 未保留时回退 "argN"（调用端按位置绑定）
            if (propName.StartsWith("arg", StringComparison.Ordinal))
            {
                propName = "arg" + index;
            }
            index++;

            var schema = TypeSchema(param.ParameterType);
            if (ap?.Description is { Length: > 0 } desc)
            {
                schema["description"] = desc;
            }
            // 值类型可空（Nullable<T>）已在 TypeSchema 合并；
            // 引用类型可空标注（string? 等）在此合并为 ["string","null"]
            if (explicitOptional && schema["type"] is JsonValue)
            {
                MarkNullable(schema);
            }
            properties[propName] = schema;

            if (markedRequired && !explicitOptional)
            {
                required.Add(propName);
            }
        }

        var root = new JsonObject { ["type"] = "object", ["properties"] = properties };
        if (required.Count > 0)
        {
            root["required"] = new JsonArray(required.Select(r => (JsonNode)r).ToArray());
        }
        return root;
    }

    /// <summary>类型 → JSON Schema（可空类型剥壳并标记可空）。</summary>
    public JsonObject TypeSchema(Type type) => TypeSchema(type, new HashSet<Type>());

    private JsonObject TypeSchema(Type type, HashSet<Type> seen)
    {
        // Nullable<T> → 剥壳 + 可空
        if (Nullable.GetUnderlyingType(type) is { } underlying)
        {
            return MarkNullable(TypeSchema(underlying, seen));
        }

        if (type == typeof(string) || type == typeof(char))
        {
            return Simple("string");
        }
        if (type == typeof(Guid))
        {
            return new JsonObject { ["type"] = "string", ["format"] = "uuid" };
        }
        if (type == typeof(DateTime) || type == typeof(DateTimeOffset))
        {
            return new JsonObject { ["type"] = "string", ["format"] = "date-time" };
        }
        if (type == typeof(byte[]))
        {
            return new JsonObject { ["type"] = "string", ["contentEncoding"] = "base64" };
        }
        if (IsIntegral(type))
        {
            return Simple("integer");
        }
        if (IsNumber(type))
        {
            return Simple("number");
        }
        if (type == typeof(bool))
        {
            return Simple("boolean");
        }
        if (type.IsEnum)
        {
            var values = new JsonArray(
                Enum.GetNames(type).Select(name => (JsonNode)name).ToArray());
            return new JsonObject { ["enum"] = values };
        }
        if (type.IsArray)
        {
            return new JsonObject
            {
                ["type"] = "array",
                ["items"] = TypeSchema(type.GetElementType()!, seen),
            };
        }
        if (IsDictionary(type))
        {
            var valueType = DictionaryValueType(type);
            return new JsonObject
            {
                ["type"] = "object",
                ["additionalProperties"] = valueType != null ? TypeSchema(valueType, seen) : new JsonObject(),
            };
        }
        if (IsCollection(type))
        {
            var itemType = CollectionItemType(type);
            var schema = new JsonObject
            {
                ["type"] = "array",
                ["items"] = itemType != null ? TypeSchema(itemType, seen) : new JsonObject(),
            };
            if (typeof(HashSet<>).IsAssignableFrom(GetRawType(type) ?? type)
                || type.Name.StartsWith("ISet", StringComparison.Ordinal)
                || type.Name.StartsWith("HashSet", StringComparison.Ordinal))
            {
                schema["uniqueItems"] = true;
            }
            return schema;
        }
        if (type == typeof(object) || type == typeof(JsonNode) || type == typeof(JsonValue))
        {
            return new JsonObject(); // 任意值
        }
        // 嵌套 POJO（含内置可枚举的类型兜底为 object）
        if (type.IsClass || type.IsValueType)
        {
            return BeanSchema(type, seen);
        }
        return new JsonObject();
    }

    private JsonObject BeanSchema(Type beanClass, HashSet<Type> seen)
    {
        var schema = new JsonObject { ["type"] = "object" };
        var properties = new JsonObject();

        if (seen.Contains(beanClass))
        {
            schema["properties"] = properties;
            return schema; // 循环引用：截断
        }
        seen.Add(beanClass);
        try
        {
            foreach (PropertyInfo prop in beanClass.GetProperties(
                         BindingFlags.Public | BindingFlags.Instance))
            {
                if (prop.GetIndexParameters().Length > 0 || !prop.CanRead)
                {
                    continue;
                }
                var ignore = prop.GetCustomAttribute<JsonIgnoreAttribute>();
                if (ignore != null && ignore.Condition == JsonIgnoreCondition.Always)
                {
                    continue;
                }
                string name = prop.GetCustomAttribute<JsonPropertyNameAttribute>()?.Name
                              ?? ToCamelCase(prop.Name);
                properties[name] = TypeSchema(prop.PropertyType, seen);
            }
        }
        finally
        {
            seen.Remove(beanClass);
        }
        schema["properties"] = properties;
        return schema;
    }

    // ------------------------------------------------------------------
    // 工具方法
    // ------------------------------------------------------------------

    private static JsonObject MarkNullable(JsonObject inner)
    {
        if (inner["type"] is JsonValue t && t.TryGetValue<string>(out var typeName))
        {
            inner["type"] = new JsonArray(typeName, "null");
        }
        else if (inner["type"] is null)
        {
            inner["nullable"] = true;
        }
        return inner;
    }

    private static bool IsNullableLike(ParameterInfo param)
    {
        // 1) Nullable<T>（值类型可空）
        if (Nullable.GetUnderlyingType(param.ParameterType) != null)
        {
            return true;
        }
        // 2) 引用类型显式可空标注（string? 等，.NET 6+ NullabilityInfoContext）
        try
        {
            var info = new NullabilityInfoContext().Create(param);
            return info.ReadState == NullabilityState.Nullable;
        }
        catch (Exception)
        {
            return false;
        }
    }

    private static bool IsIntegral(Type type) => type == typeof(byte) || type == typeof(sbyte)
        || type == typeof(short) || type == typeof(ushort) || type == typeof(int)
        || type == typeof(uint) || type == typeof(long) || type == typeof(ulong);

    private static bool IsNumber(Type type) => type == typeof(float) || type == typeof(double)
        || type == typeof(decimal);

    private static bool IsDictionary(Type type)
    {
        if (!type.IsGenericType)
        {
            return false;
        }
        foreach (Type iface in type.GetInterfaces())
        {
            if (iface.IsGenericType &&
                (iface.GetGenericTypeDefinition() == typeof(IDictionary<,>)
                 || iface.GetGenericTypeDefinition() == typeof(IReadOnlyDictionary<,>)))
            {
                return true;
            }
        }
        var def = type.GetGenericTypeDefinition();
        return def == typeof(IDictionary<,>) || def == typeof(IReadOnlyDictionary<,>)
               || def == typeof(Dictionary<,>);
    }

    private static bool IsCollection(Type type)
    {
        if (!type.IsGenericType)
        {
            return false;
        }
        if (IsDictionary(type))
        {
            return false;
        }
        foreach (Type iface in type.GetInterfaces())
        {
            // 注意：开放泛型接口（IEnumerable<>）对 List<string> 等封闭类型
            // IsAssignableFrom 返回 false，必须扫描已实现接口
            if (iface.IsGenericType && iface.GetGenericTypeDefinition() == typeof(IEnumerable<>))
            {
                return true;
            }
        }
        // 参数本身就是 IEnumerable<T> 接口
        return type.GetGenericTypeDefinition() == typeof(IEnumerable<>);
    }

    private static Type? DictionaryValueType(Type type)
    {
        foreach (Type iface in type.GetInterfaces())
        {
            if (iface.IsGenericType &&
                (iface.GetGenericTypeDefinition() == typeof(IDictionary<,>)
                 || iface.GetGenericTypeDefinition() == typeof(IReadOnlyDictionary<,>)))
            {
                return iface.GetGenericArguments()[1];
            }
        }
        if (type.IsGenericType)
        {
            var def = type.GetGenericTypeDefinition();
            if (def == typeof(Dictionary<,>) || def == typeof(IDictionary<,>)
                || def == typeof(IReadOnlyDictionary<,>))
            {
                return type.GetGenericArguments()[1];
            }
        }
        return null;
    }

    private static Type? CollectionItemType(Type type)
    {
        foreach (Type iface in type.GetInterfaces())
        {
            if (iface.IsGenericType && iface.GetGenericTypeDefinition() == typeof(IEnumerable<>))
            {
                return iface.GetGenericArguments()[0];
            }
        }
        if (type.IsGenericType && type.GetGenericTypeDefinition() is { } def
                                && def != typeof(IDictionary<,>)
                                && def != typeof(IReadOnlyDictionary<,>))
        {
            return type.GetGenericArguments()[0];
        }
        return null;
    }

    private static Type? GetRawType(Type type) => type.IsGenericType ? type.GetGenericTypeDefinition() : null;

    private static JsonObject Simple(string type) => new() { ["type"] = type };

    private static string ToCamelCase(string name)
        => name.Length > 1 && char.IsUpper(name[0])
            ? char.ToLowerInvariant(name[0]) + name[1..]
            : name;
}